xdsl/dialects/riscv.py

```diff
@@ -331,6 +333,106 @@ class RISCVOp(Operation, ABC):
     def assembly_line(self) -> str | None:
         raise NotImplementedError()
 
+    @classmethod
+    def parse(cls, parser: Parser) -> Self:
+        return cls.create(
+            operands=cls.parse_operands(parser),
+            successors=cls.parse_successors(parser),
+            regions=cls.parse_regions(parser),
+            attributes=cls.parse_attributes(parser),
+            result_types=cls.parse_result_types(parser)[1],
+        )
+
+    @classmethod
+    def parse_operands(cls, parser: Parser) -> Sequence[SSAValue]:
+        operands = parser.parse_comma_separated_list(
+            parser.Delimiter.PAREN, parser.parse_operand
+        )
+        return operands
+
+    @classmethod
+    def parse_successors(cls, parser: Parser) -> Sequence[Block]:
+        return parser.parse_optional_successors() or []
+
+    @classmethod
+    def parse_result_types(
+        cls, parser: Parser
+    ) -> tuple[Sequence[Attribute], Sequence[Attribute]]:
+        parser.parse_punctuation(":")
+        args = parser.parse_comma_separated_list(
+            Parser.Delimiter.PAREN,
+            parser.parse_type,
+        )
+        parser.parse_punctuation("->")
+        if parser.parse_optional_punctuation("(") is None:
+            returns = [parser.parse_type()]
+        else:
+            returns = list[Attribute]()
+            if parser.parse_optional_punctuation(")") is None:
+                returns.append(parser.parse_type())
+                while parser.parse_optional_punctuation(",") is not None:
+                    returns.append(parser.parse_type())
+                parser.parse_punctuation(")")
+        return args, returns
```

Comment:
I am curious why this code is not available already. AFAIU types should already be parsed by the generic parser, no? @kingiler, can you explain why the generic code in the parser won't work?

Reply:
Because it is in a private method `Parser.parse_function_type` and `pyright` is not happy about accessing it so I just copy and paste for now. Making a PR to make it a public function.

Reply:
Use updated parser function in newest commit.